burn-core/src/record/primitive.rs

```diff
@@ -183,3 +189,79 @@ primitive!(i64);
 primitive!(i32);
 primitive!(i16);
 primitive!(i8);
+
+/// A wrapper around an array of size N, so that it can be serialized and deserialized
+/// using serde.
+///
+/// The reason why we wrap the array in a struct is because serde does not support
+/// deserializing arrays of variable size,
+/// see [serde/issues/1937](https://github.com/serde-rs/serde/issues/1937)
+/// for backward compatibility reasons. Serde APIs were created before const generics.
+pub struct Array<const N: usize, T>([T; N]);
+
+impl<T: Serialize, const N: usize> Serialize for Array<N, T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut seq = serializer.serialize_tuple(self.0.len())?;
+        for element in &self.0 {
+            seq.serialize_element(element)?;
+        }
+        seq.end()
+    }
+}
+
+impl<'de, T, const N: usize> Deserialize<'de> for Array<N, T>
+where
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct ArrayVisitor<T, const N: usize> {
+            marker: PhantomData<T>,
+        }
+
+        impl<'de, T, const N: usize> Visitor<'de> for ArrayVisitor<T, N>
+        where
+            T: Deserialize<'de>,
+        {
+            type Value = Array<N, T>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a fixed size array")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let mut items = vec![];
+
+                for i in 0..N {
+                    let item = seq
+                        .next_element()?
+                        .ok_or_else(|| Error::invalid_length(i, &self))?;
+                    items.push(item);
+                }
```

Comment:
If you mean to iterate through a ranch and use map, then it's probably longer and complicated. Please suggest how it might look like, I am open to change.